blocks/rich-text/inline-blocks.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, Fragment, compose } from '@wordpress/element';
+import { withSelect, withDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import InlineInsertionPoint from './inline-insertion-point';
+import MediaUpload from '../media-upload';
+
+class InlineBlocks extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.insert = this.insert.bind( this );
+		this.onSelectMedia = this.onSelectMedia.bind( this );
+		this.openMediaLibrary = this.openMediaLibrary.bind( this );
+		this.closeMediaLibrary = this.closeMediaLibrary.bind( this );
+		this.state = { mediaLibraryOpen: false };
+	}
+
+	componentDidMount() {
+		// When moving between two different RichText with the keyboard, we need to
+		// make sure `setInsertAvailable` is called after `setInsertUnavailable`
+		// from previous RichText so that editor state is correct
+		setTimeout( this.props.setInsertAvailable );
+	}
+
+	componentDidUpdate( prevProps ) {
+		if (
+			this.props.inlineBlockForInsert &&
+			! prevProps.inlineBlockForInsert
+		) {
+			this.insert();
+		}
+	}
+
+	componentWillUnmount() {
+		this.props.setInsertUnavailable();
+	}
+
+	insert() {
+		const {
+			inlineBlockForInsert,
+			completeInsert,
+			editor,
+		} = this.props;
+
+		if ( inlineBlockForInsert.type === 'image' ) {
+			this.openMediaLibrary();
```

Comment:
I think this is still a work in progress, but just noting for now that this does not belong in `RichText`, this should move to the block registration file.

Reply:
@iseulde Starting to work on this, along with having the inline block return a React element as you suggested in https://github.com/WordPress/gutenberg/pull/5794#discussion_r185042816. Do you think this should be limited to a single React element or could it be a full component with lifecycle methods? For the inline image it seems like it may need to be a component in order to include the MediaUpload logic. But I'm not sure if TinyMCE content can include React components. I'll look at this more tomorrow, but would appreciate any thoughts you or others have.

Reply:
Hm, I'm not quite sure. RichText shouldn't be involved here until the React Element/HTML is ready to be inserted. Maybe it can be mounted in the inserter?